### PR TITLE
Add validate generator

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

This PR adds a `validate` generator in the controller-tools-based generators that allow maintaining the K8S API. 
This generator validates the consistency and correctness of the K8S API GO source code, against
pattern, especially the unions. 
If something is wrong, like a union without an optional discriminator, or a discriminator without the correct kubebuilder enumeration annotation, then an error is issued with the precise place in the source file where the problem is.

### How to review this PR?

It should be easier to review this PR commit-by-commit.

### What issues does this PR fix or reference?

This PR doesn't relate to any issue.

### Is your PR tested? Consider putting some instruction how to test your changes

Yes.

### CODE REVIEW

File: generator/validate/gen.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
+package validate
```
Comment (JPinkney) [inline generator/validate/gen.go:1]: Should we have license header for this?

Comment (amisevsk) [reply generator/validate/gen.go:1]: https://github.com/devfile/api/pull/137 :)

**Review by amisevsk — COMMENTED** [review]
I have a few comments and I think some refactoring is necessary for clarity (but don't yet have a clean suggestion for that).

My suggestions about reducing nesting are in PR https://github.com/davidfestal/api/pull/1 if you want to take a look.

File: generator/validate/gen.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+						root.AddError(loader.ErrFromNode(fmt.Errorf(
+							"In union `%v` the union discriminator `%v` should have the `+optional` comment marker",
+							typeToCheck.Name,
+							unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+					}
+					if !strings.Contains(unionDiscriminatorField.Tag.Get("json"), ",omitempty") {
+						root.AddError(loader.ErrFromNode(fmt.Errorf(
+							"In union `%v` the union discriminator `%v` should contain the `omitempty` option in its `json` tag, since it is expected to be an optional field",
+							typeToCheck.Name,
+							unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply generator/validate/gen.go:87]: Thinking about it more, though, maybe not since so much context is needed.

File: generator/validate/gen.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+							}
+							root.AddError(loader.ErrFromNode(fmt.Errorf(
+								"In union `%v` the union discriminator `%v` should only have the kubebuilder marker that defines the allowed union values as an Enum (kubebuilder:validation:Enum=....)",
+								typeToCheck.Name,
+								unionDiscriminatorField.Name), node))
+						}
+						return nil
+					}
+
+					enumMarkerPtr := getEnumValidation(unionDiscriminatorField.Markers, unionDiscriminatorField.RawField)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply generator/validate/gen.go:119]: To add here, I think in the latter case we should force the enum annotation to be defined on the `discriminatorType` rather than in the struct field. I can get past validation by moving the `enum` annotation into the Component struct, for example.

File: generator/validate/gen.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		})
+
+		root.NeedTypesInfo()
+
+		packageTypes := map[string]*markers.TypeInfo{}
+		if err := markers.EachType(ctx.Collector, root, func(info *markers.TypeInfo) {
+			packageTypes[info.Name] = info
+		}); err != nil {
+			root.AddError(err)
+			return nil
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go:50]: Only adding the error to root allows the other generators to run (when several generators are run sequentially in one command line). The errors are output at the end of the process.
When returning the error in the generator itself, I assume it would stop the command line here.

This doesn't really make a difference in our case since we use a single generator in each command line run.

File: devfile.api.code-workspace [hunk 5]
```diff
@@ -89,13 +89,13 @@
                 "cwd": "${workspaceFolder:api}"
             },
             {
-                "name": "Launch All Generators",
+                "name": "Launch Validate Generator",
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply devfile.api.code-workspace:92]: No, this doesn't work properly in our case due to the fact that some generator steps must have access to the files generators by previous generator.

File: generator/validate/gen.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+						}
+						return nil
+					}
+
+					enumMarkerPtr := getEnumValidation(unionDiscriminatorField.Markers, unionDiscriminatorField.RawField)
+					var enumAstNode ast.Node = unionDiscriminatorField.RawField
+
+					if _, isBasic := discriminatorTypeRef.Type.(*types.Basic); !isBasic {
+						if typeNamed, isNamed := discriminatorTypeRef.Type.(*types.Named); isNamed {
+							enumTypeInfo, isInSamePackage := packageTypes[typeNamed.Obj().Name()]
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go:124]: The logic was the following: all the types defined in the current package are in the `packageTypes` map.
So if the name of the discriminator type is not found in this map, that means that the type of the discriminator is not in the same package, which explain the error.
Does it make sense ?

File: generator/validate/gen.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+					underlyingType, underlyingTypeIsBasic := discriminatorTypeRef.Type.Underlying().(*types.Basic)
+					if !underlyingTypeIsBasic || underlyingType.Kind() != types.String {
+						root.AddError(wrongTypeError)
+					}
+
+					getEnumValidation := func(markers markers.MarkerValues, node ast.Node) *crdmarkers.Enum {
+						enumMarkerIf := markers.Get("kubebuilder:validation:Enum")
+						if enumMarkerIf != nil {
+							if enumMarker, isEnum := enumMarkerIf.(crdmarkers.Enum); isEnum {
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go:108]: At start I thought that we could possibly have several types of unions => several types of structs. 
But that's not the case, so I'll simplify the code here.

File: generator/validate/gen.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+					if !typeFound {
+						root.AddError(wrongTypeError)
+					}
+
+					underlyingType, underlyingTypeIsBasic := discriminatorTypeRef.Type.Underlying().(*types.Basic)
+					if !underlyingTypeIsBasic || underlyingType.Kind() != types.String {
+						root.AddError(wrongTypeError)
+					}
+
+					getEnumValidation := func(markers markers.MarkerValues, node ast.Node) *crdmarkers.Enum {
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go:105]: I'll rename it yes

File: generator/validate/gen.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply generator/validate/gen.go:124]: Makes sense :+1:

File: generator/validate/gen.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply generator/validate/gen.go:119]: The problem is that the checks are interleaved, making it hard to follow:
1. Get enum marker from struct field
2. Check type of struct field; if it's not `string`, then
    * a. Check that type def for struct field's type is in the same package
    * b. Get enum marker for type def
    * c. Check if type def enum marker from b is `nil`; if it is not, then
        * Check whether struct field enum marker from 1. is `nil`; if it is not add an error for two enum markers
        * Otherwise, overwrite enum marker from step 1 with the type def enum marker
3. Check if enum marker is `nil` and add error if it is (not caring if it was set in step 1 or step c.2)
4. Check enum marker's contents to match expected

In the first case (type of struct field is `string`), the check is
1. Get enum marker, ensure not `nil`
2. Check enum marker contents

In the second case (struct field is a named type), the check is (or could be simplified to)
1. Check no enum marker on struct field
2. Check that there is an enum marker on named type
3. Check contents of enum marker

[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go]: Fixed in commit https://github.com/devfile/api/pull/142/commits/5fb3883fc4072194ad8a9a6e1b2f4ee9f5dba691

File: generator/validate/gen.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go]: Fixed in commit https://github.com/devfile/api/pull/142/commits/5fb3883fc4072194ad8a9a6e1b2f4ee9f5dba691

File: generator/validate/gen.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go]: Fixed in commit https://github.com/devfile/api/pull/142/commits/5fb3883fc4072194ad8a9a6e1b2f4ee9f5dba691

File: generator/validate/gen.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+		packageTypes := map[string]*markers.TypeInfo{}
+		if err := markers.EachType(ctx.Collector, root, func(info *markers.TypeInfo) {
+			packageTypes[info.Name] = info
+		}); err != nil {
+			root.AddError(err)
+			return nil
+		}
+
+		for _, typeToCheck := range packageTypes {
+			if typeToCheck.Markers.Get(genutils.UnionMarker.Name) != nil {
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go]: Fixed in commit https://github.com/devfile/api/pull/142/commits/5fb3883fc4072194ad8a9a6e1b2f4ee9f5dba691 by extract the union checking in a function.

File: generator/validate/gen.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+					} else {
+						unionMembers = append(unionMembers, unionField.Name)
+					}
+				}
+				if unionDiscriminatorField == nil {
+					root.AddError(loader.ErrFromNode(fmt.Errorf(
+						"Union `%v` should have 1 union discriminator. See here for details: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#proposal",
+						typeToCheck.Name), typeToCheck.RawSpec))
+				} else {
+					if unionDiscriminatorField.Markers.Get("optional") == nil {
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply generator/validate/gen.go]: Fixed in commit https://github.com/devfile/api/pull/142/commits/5fb3883fc4072194ad8a9a6e1b2f4ee9f5dba691 by extract the union checking in a function.

**Review by amisevsk — APPROVED** [review]
LGTM, thanks @davidfestal!

File: generator/validate/gen.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+		return
+	}
+
+	unionMembers := []string{}
+	var unionDiscriminatorField *markers.FieldInfo
+	for index, unionField := range typeToCheck.Fields {
+		if unionField.Markers.Get(genutils.UnionDiscriminatorMarker.Name) != nil {
+			if unionDiscriminatorField != nil {
+				root.AddError(loader.ErrFromNode(fmt.Errorf(
+					"Union `%v` should have only 1 union discriminator, but has 2: `%v` and `%v`",
```
Comment (sleshchenko) [inline generator/validate/gen.go:73]: that's not critical if we can go go-style we would avoid warnings, so Go has convention not to capitalize error messages https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: generator/validate/gen.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+				"In union `%v` the union discriminator `%v` should have a string-based type defined in the same package",
+				typeToCheck.Name,
+				unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+		}
+	}
+
+	if fieldEnumMarkerPtr != nil && typeEnumMarkerPtr != nil {
+		root.AddError(loader.ErrFromNode(fmt.Errorf(
+			"Type `%v` should not define a `kubebuilder:validation:Enum` annotation, since it is already defined in the discriminator `%v` of union `%v`",
+			enumTypeInfo.Name,
```
Comment (sleshchenko) [inline generator/validate/gen.go:148]: enumTypeInfo is nullable here. I assume return is missing after line 141

File: generator/validate/gen.go [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+					typeToCheck.Name,
+					unionDiscriminatorField.Name), node))
+				break
+			}
+		}
+	}
+	if fieldEnumMarkerPtr != nil {
+		checkEnumValues(fieldEnumMarkerPtr, unionDiscriminatorField.RawField)
+	} else {
+		checkEnumValues(typeEnumMarkerPtr, enumTypeInfo.RawSpec)
```
Comment (sleshchenko) [inline generator/validate/gen.go:176]: enumTypeInfo is nullable here

**Review by sleshchenko — APPROVED** [review]

File: generator/validate/gen.go (hunk 12 above)
Comment (davidfestal) [reply generator/validate/gen.go:148]: Well, not really. In this case we know that `typeEnumMarkerPtr != nil`, and the case when `typeEnumMarkerPtr` can be set to a non-nil pointer, is precisely when `enumTypeInfo != nil`.

File: generator/validate/gen.go (hunk 13 above)
Comment (davidfestal) [reply generator/validate/gen.go:176]: Same answer here : `typeEnumMarkerPtr != nil` => `enumTypeInfo != nil`

File: generator/validate/gen.go (hunk 11 above)
Comment (davidfestal) [reply generator/validate/gen.go]: Fixed in commit https://github.com/devfile/api/pull/142/commits/125c893a7a747127686c3ec73226f3d1eb02c533
